p2p/cuprate-p2p/src/block_downloader.rs

```diff
@@ -0,0 +1,1181 @@
+
+    /// The current inflight requests.
+    ///
+    /// This a map of batch start heights to block ids and related information of the batch.
+    inflight_requests: BTreeMap<u64, BlocksToRetrieve<N>>,
+
+    /// A queue of ready batches.
+    ready_batches: BinaryHeap<ReadyQueueBatch>,
+    /// The size, in bytes, of all the batches in `ready_batches`.
+    ready_batches_size: usize,
+
+    /// A queue of failed batches start height's that should be retried.
+    ///
+    /// Wrapped in [`Reverse`] so we prioritize early batches.
+    failed_batches: BinaryHeap<Reverse<u64>>,
+
+    /// The [`BufferAppender`] that gives blocks to Cuprate.
+    buffer_appender: BufferAppender<BlockBatch>,
+
+    /// The [`BlockDownloaderConfig`].
+    config: BlockDownloaderConfig,
+}
+
+impl<N: NetworkZone, S, C> BlockDownloader<N, S, C>
+where
+    S: PeerSyncSvc<N> + Clone,
+    C: Service<ChainSvcRequest, Response = ChainSvcResponse, Error = tower::BoxError>
+        + Send
+        + 'static,
+    C::Future: Send + 'static,
+{
+    /// Creates a new [`BlockDownloader`]
+    fn new(
+        client_pool: Arc<ClientPool<N>>,
+
+        peer_sync_svc: S,
+        our_chain_svc: C,
+        buffer_appender: BufferAppender<BlockBatch>,
+
+        config: BlockDownloaderConfig,
+    ) -> Self {
+        BlockDownloader {
+            client_pool,
+            pending_peers: BTreeMap::new(),
+            peer_sync_svc,
+            our_chain_svc,
+            amount_of_blocks_to_request: config.initial_batch_size,
+            amount_of_blocks_to_request_updated_at: 0,
+            amount_of_empty_chain_entries: 0,
+            block_download_tasks: JoinSet::new(),
+            chain_entry_task: JoinSet::new(),
+            inflight_requests: BTreeMap::new(),
+            ready_batches: BinaryHeap::new(),
+            ready_batches_size: 0,
+            failed_batches: BinaryHeap::new(),
+            buffer_appender,
+            config,
+        }
+    }
+
+    /// Checks if we can make use of any peers that are currently pending requests.
+    async fn check_pending_peers(&mut self, chain_tracker: &mut ChainTracker<N>) {
+        tracing::debug!("Checking if we can give any work to pending peers.");
+
+        // HACK: The borrow checker doesn't like the following code if we don't do this.
+        let mut pending_peers = mem::take(&mut self.pending_peers);
+
+        for (_, peers) in pending_peers.iter_mut() {
+            while let Some(peer) = peers.pop() {
+                if peer.info.handle.is_closed() {
+                    // Peer has disconnected, drop it.
+                    continue;
+                }
+
+                if let Some(peer) = self.try_handle_free_client(chain_tracker, peer).await {
+                    // This peer is ok however it does not have the data we currently need, this will only happen
+                    // because of it's pruning seed so just skip over all peers with this pruning seed.
```

Comment:
```suggestion
                    // This peer is ok however it does not have the data we currently need, this will only happen
                    // because of its pruning seed so just skip over all peers with this pruning seed.
```